### PR TITLE
[torch.compile] Dynamic fp8 + rms_norm fusion

### PR DESCRIPTION
**Pull request by ProExpertProg**
This PR cleans up the fusion pass to make it easier to add other multi-output patterns. Then it adds dynamic fp8 rmsnorm fusion.

### CODE REVIEW

File: csrc/quantization/fused_kernels/fused_layernorm_dynamic_per_token_quant.cu [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+      &token_scale, scales, input, weight, rms, scale_ub, min_scaling_factor,
+      hidden_size, residual);
+
+  // RMS Norm + Quant
+  if constexpr (std::is_same_v<scalar_out_t, int8_t>) {
+    vllm::vectorized::norm_and_quant<scalar_t, scalar_out_t, true,
+                                     has_residual>(
+        out, input, weight, rms, 1.0f / token_scale, hidden_size, residual);
+  } else {
+    // FP8 - Do not invert s_token_scale for exact match with FBGemm
```
Comment (varun-sundar-rabindranath) [inline csrc/quantization/fused_kernels/fused_layernorm_dynamic_per_token_quant.cu:39]: `s_token_scale` -> `token_scale`

File: csrc/quantization/fused_kernels/layernorm_utils.cuh [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+  for (int32_t i = threadIdx.x; i < hidden_size; i += blockDim.x) {
+    float x = static_cast<float>(input[token_offset + i]);
+    if constexpr (has_residual) {
+      x += static_cast<float>(residual[token_offset + i]);
+    }
+
+    ss += x * x;
+  }
+
+  using BlockReduce = cub::BlockReduce<float, 1024>;
```
Comment (varun-sundar-rabindranath) [inline csrc/quantization/fused_kernels/layernorm_utils.cuh:36]: the block_dim.x is defined as 
```
 dim3 block(std::min(hidden_size, 1024));
```
is it safe doing `cub::BlockReduce<float,1024>` when block_dim.x is < 1024 ?

File: tests/kernels/test_fused_quant_layernorm.py [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+import torch
+
+import vllm._custom_ops as ops
+from vllm.model_executor.layers.layernorm import RMSNorm
+
+DTYPES = [torch.bfloat16, torch.float]
+QUANT_DTYPES = [torch.int8, torch.float8_e4m3fn]
+NUM_TOKENS = [1, 7, 83, 2048, 4096]  # Arbitrary values for testing
+HIDDEN_SIZES = [1, 2, 3, 4, 16, 64, 67, 768, 2048, 5120, 5137, 8192,
+                8193]  # Arbitrary values for testing
```
Comment (varun-sundar-rabindranath) [inline tests/kernels/test_fused_quant_layernorm.py:13]: We can probably reduce the hidden_sizes to [1, 3, 4, 16, 64, 2048, 5120, 5137] + the vectorization edge-cases to save test times.

File: vllm/_custom_ops.py [hunk 4]
```diff
@@ -22,6 +22,7 @@
 supports_moe_ops = False
 with contextlib.suppress(ImportError):
     import vllm._moe_C  # noqa: F401
+
```
Comment (varun-sundar-rabindranath) [inline vllm/_custom_ops.py:25]: nit: whitespace changes, here and below

**Comment by varun-sundar-rabindranath** [comment]
Reviewed the kernel files and kernel tests. Left some minor comments. LGTM otherwise.
